tree-sitter/src/bridge.c

```diff
@@ -95,10 +93,33 @@ void ts_tree_cursor_reset_p(TSTreeCursor *cursor, TSNode *node) {
   ts_tree_cursor_reset(cursor, *node);
 }
 
-/* FIXME: copy out a Node rather than a TSNode */
-bool ts_tree_cursor_current_node_p(const TSTreeCursor *cursor, TSNode *outNode) {
+bool ts_tree_cursor_current_node_p(const TSTreeCursor *cursor, Node *outNode) {
   assert(cursor != NULL);
   assert(outNode != NULL);
-  *outNode = ts_tree_cursor_current_node(cursor);
-  return !ts_node_is_null(*outNode);
+  TSNode tsNode = ts_tree_cursor_current_node(cursor);
+  if (!ts_node_is_null(tsNode)) {
+    ts_node_poke(ts_tree_cursor_current_field_name(cursor), tsNode, outNode);
+  }
+  return false;
+}
+
+
+uint32_t ts_tree_cursor_copy_child_nodes(TSTreeCursor *cursor, Node *outChildNodes) {
+  assert(cursor != NULL);
+  assert(outChildNodes != NULL);
+  uint32_t count = 0;
+
+  if (ts_tree_cursor_goto_first_child(cursor)) {
+    do {
+      TSNode current = ts_tree_cursor_current_node(cursor);
+      const char *fieldName = ts_tree_cursor_current_field_name(cursor);
+      if (fieldName || (ts_node_is_named(current) && !ts_node_is_extra(current))) {
+        ts_node_poke(fieldName, current, outChildNodes);
+        count++;
+        outChildNodes++;
```

Comment:
I don’t think this is right? You’re incrementing the pointer here, so this is a no-op and will be optimized away since you never read from it subsequently. I think you want to delete this line, and then after this loop ends (right before the `return`), write

`*outChildNodes = count`.

Reply:
You could also write `++*outChildNodes`, which is a little harder to read, but in my testing generates tighter assembly (one fewer jump);